docs: add advanced section for module authors (#2950)
nuxt/framework: docs/content/3.docs/4.advanced/3.kit.md

# Nuxt Kit

> > Nuxt kit provide composable utilities to making interaction with [/docs/advanced/hooks](Nuxt Hooks) and [Nuxt Builder Core](/docs/advanced/nuxt#the-nuxt-interface) and developing [Nuxt Modules](/docs/advanced/modules) super easy!

Reviewer: ```suggestion
> Nuxt Kit provides composable utilities to make interacting with [/docs/advanced/hooks](Nuxt Hooks) and [Nuxt Builder Core](/docs/advanced/nuxt#the-nuxt-interface) and developing [Nuxt Modules](/docs/advanced/modules) super easy!
```

## Usage

### Install dependency

You can install the latest nuxt kit by adding it to `dependencies` of `package.json`. However, please consider always explicitly installing the `@nuxt/kit` package even if it is already installed by nuxt.

```json [package.json]
{
  "dependencies": {
    "@nuxt/kit": "npm:@nuxt/kit-edge@latest"
  }
}
```

### Import kit utilities

```js [test.mjs]
import { useNuxt } from '@nuxt/kit'
```

::alert{type="warning"}
Nuxt kit utilities are only available for modules and not meant to be imported in runtime (components, vue composables, pages, plugins, or server routes)
::

Nuxt kit, is an [esm-only package](/concepts/esm) meaning you **cannot** `require('@nuxt/kit')`. As a workaround, we can use dynamic import to use it in the CommonJS context:

```js [test.cjs]
// This does NOT work!
// const kit = require('@nuxt/kit')
async function main() {
  const kit = await import('@nuxt/kit')
}
main()
```

### Use kit utilities

Nuxt context is globally available using [unjs/unctx](https://github.com/unjs/unctx). Because of this, most kit composables do not need to have `nuxt` instances explicitly passed.

```js [test.mjs]

> import { getNuxtVersion }

Reviewer: ```suggestion
import { getNuxtVersion } from '@nuxt/kit'
```

> import { getNuxtVersion() }

Reviewer: ```suggestion
import { getNuxtVersion }
```
